### ISSUE TITLE
Programmatic setting of config

### ISSUE THREAD

**Issue by digitalsadhu** [podium-lib/labs#4]
Currently, config is loaded early in the process and the first opportunity developers will get to set config values programmatically is in `server.js`. For example:

```js
export default async function server(app, { config }) {    
    if (!app.config.get('assets.development')) {
        app.config.set('assets.base', 'https://assets.finn.no/etc');
    }
}
```

This seems like a fairly intuitive solution to reach for however there's an issue. Since config is loaded early and the Fastify plugin that is server.js is loaded relatively late, many of the config values that have been set, have also been used and changing will have no effect. Eg. The `app.name` value has been used in the new Podlet({}) call at this point.

My instinct is that we should say "don't set config values in server.js" since there will always be some values that will not be alterable In server.js even if some were. (values passed to new Podlet have to have been set before server.js where as asset.base could be refactored so that its possible to set it that late perhaps) It feels much cleaner to me to say, don't do it, than to try to enable some values to be settable in server.js while others cannot be. 

Continuing with the issue... The obvious solution would seem to be to just set values in the config folder but there's a problem there too. Currently we only support json files where no logic is possible. It wouldn't be possible to read in eik.json, build the asset url and then set it in config for example.

Currently we support config files of the following types

```
/config
  /common.json
  /domains
    /www.finn.no
      /production.config.json
```

The first solution that comes to mind is that we support config files defined in js as well as json.

Eg. 
```json5
// /config/common.json
{
  "app": {
    "name": "can only be static"
  }
}
```
could be written as
```js
// /config/common.js
import { name } from "../package.json" assert { type: "json" }
export default {
  app: {
    name,
  }
}
```

Thoughts?

**Comment by trygve-lie** [comment]
> My instinct is that we should say "don't set config values in server.js"

That is also my instinct on config. Config for a server is something the server loads as one of the first things it need to proceed. Config holds static stuff the app will need to start (http port number etc) and continue to run. 

I actually don't think it should be possible to override one config config property after the config as been resolved and validated. Doing so can create a number of un-intuitive scenarios as I see it. The way that convict works is that it will first load the schema, then it will load config files (json) for the environment the service is set to run in and then it will possibly override a property with environment variables or command line arguments if the property is configured to take these into account. 

Iow; its a common pattern to ex in a container override something with environment variables. If a user then has a config schema and overrides it with an environment variable the user will probably expect the environment variable to take precedence. But if later down the road in the app, there is something which just sets a config property which overrides this again it would be very un-intuitive behavior.

So I am very much leaning against not setting config properties outside of the bootstrapping process of loading config.

But in the context of using js as config; the user can here define a schema and in there its possible to do more programmatic logic. How is that not covering our use cases?

**Comment by digitalsadhu** [comment]
Makes sense. The example I have in mind that is tricky to solve is this:

Internally we hang the assets base path off a config variable called `assets.base`. 
We want to dynamically generate this using the eik.json file values when we override it.
We can't just define a new schema value for this since we'd still need to override base somewhere.
It would be possible to do this using something like this in a file that is loaded early, like schema.js

```js
const eik = JSON.parse(
  await readFile(new URL('./eik.json', import.meta.url), {
    encoding: 'utf8',
  }));
  process.env.ASSETS_BASE = join(eik.server, 'pkg', eik.name, eik.version);
}
```

but feels pretty forced. And even then it doesn't do what the Eik node client does with switching paths between dev and prod.
